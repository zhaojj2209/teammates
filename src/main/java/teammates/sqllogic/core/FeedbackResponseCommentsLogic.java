package teammates.sqllogic.core;

import java.util.UUID;

import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.storage.sqlapi.FeedbackResponseCommentsDb;
import teammates.storage.sqlentity.FeedbackResponseComment;
import teammates.ui.request.FeedbackResponseCommentUpdateRequest;

/**
 * Handles operations related to feedback response comments.
 *
 * @see FeedbackResponseComment
 * @see FeedbackResponseCommentsDb
 */
public final class FeedbackResponseCommentsLogic {

    private static final FeedbackResponseCommentsLogic instance = new FeedbackResponseCommentsLogic();
    private FeedbackResponseCommentsDb frcDb;

    private FeedbackResponseCommentsLogic() {
        // prevent initialization
    }

    public static FeedbackResponseCommentsLogic inst() {
        return instance;
    }

    /**
     * Initialize dependencies for {@code FeedbackResponseCommentsLogic}.
     */
    void initLogicDependencies(FeedbackResponseCommentsDb frcDb) {
        this.frcDb = frcDb;
    }

    /**
     * Gets an feedback response comment by feedback response comment id.
     * @param id of feedback response comment.
     * @return the specified feedback response comment.
     */
    public FeedbackResponseComment getFeedbackResponseComment(Long id) {
        return frcDb.getFeedbackResponseComment(id);
    }

    /**
     * Gets the comment associated with the response.
     */
    public FeedbackResponseComment getFeedbackResponseCommentForResponseFromParticipant(
            UUID feedbackResponseId) {
        return frcDb.getFeedbackResponseCommentForResponseFromParticipant(feedbackResponseId);
    }

    /**
     * Creates a feedback response comment.
     * @throws EntityAlreadyExistsException if the comment alreadty exists
     * @throws InvalidParametersException if the comment is invalid
     */
    public FeedbackResponseComment createFeedbackResponseComment(FeedbackResponseComment frc)
            throws InvalidParametersException, EntityAlreadyExistsException {
        return frcDb.createFeedbackResponseComment(frc);
    }

    /**
     * Deletes a feedbackResponseComment.
     */
    public void deleteFeedbackResponseComment(Long frcId) {
        frcDb.deleteFeedbackResponseComment(frcId);
    }

    /**
     * Updates a feedback response comment.
     * @throws EntityDoesNotExistException if the comment does not exist
     */
    public FeedbackResponseComment updateFeedbackResponseComment(Long frcId,
            FeedbackResponseCommentUpdateRequest updateRequest, String updaterEmail)
            throws EntityDoesNotExistException {
        FeedbackResponseComment comment = frcDb.getFeedbackResponseComment(frcId);
        if (comment == null) {
            throw new EntityDoesNotExistException("Trying to update a feedback response comment that does not exist.");
        }

        comment.setCommentText(updateRequest.getCommentText());
        comment.setShowCommentTo(updateRequest.getShowCommentTo());
        comment.setShowGiverNameTo(updateRequest.getShowGiverNameTo());
        comment.setLastEditorEmail(updaterEmail);

        return comment;
    }

    /**
     * Updates all email fields of feedback response comments.
     */
    public void updateFeedbackResponseCommentsEmails(String courseId, String oldEmail, String updatedEmail) {
        frcDb.updateGiverEmailOfFeedbackResponseComments(courseId, oldEmail, updatedEmail);
        frcDb.updateLastEditorEmailOfFeedbackResponseComments(courseId, oldEmail, updatedEmail);
    }

}
