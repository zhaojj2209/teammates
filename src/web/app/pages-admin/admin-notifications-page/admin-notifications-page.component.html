<h1>Notifications</h1>
<p>This page allows you to create and manage notifications which are shown to respective target users when they access TEAMMATES.</p>
<div id="notifications-timezone">
  <p>All dates are displayed in {{ guessTimezone }} time.</p>
</div>

<div>
  <button id="btn-add-notification" class="btn btn-success" (click)="isNotificationEditFormExpanded = true"
          [disabled]="isNotificationEditFormExpanded || isNotificationLoading">
          <span *ngIf="currentNotificationEditFormMode === NotificationEditFormMode.ADD else editbtn"><i class="fas fa-plus"></i> Add New Notification</span>
          <ng-template #editbtn><i class="fas fa-cog"></i> Edit Existing Notification</ng-template>
  </button>
</div>

<div class="margin-top-30px" *ngIf="isNotificationEditFormExpanded && !isNotificationLoading" @collapseAnim>
  <tm-notification-edit-form [formMode]="currentNotificationEditFormMode"
                             [(model)]="notificationEditFormModel"
                             [guessTimezone]="guessTimezone"
                             (addNewNotificationEvent)="addNewNotificationHandler()"
                             (editExistingNotificationEvent)="editExistingNotificationHandler()"
                             (cancelEditingNotificationEvent)="initNotificationEditFormModel()"></tm-notification-edit-form>
</div>

<tm-loading-retry [shouldShowRetry]="hasNotificationLoadingFailed" [message]="'Failed to load data'"
                  (retryEvent)="retryLoadingAllData()">
  <div class="margin-top-30px">
    <div *tmIsLoading="isNotificationLoading">
      <div class="alert alert-warning margin-top-30px" role="alert" *ngIf="notificationsTableRowModels.length === 0 && !isNotificationLoading">
        You have not created any notifications yet. Use the form above to create a notification.
      </div>

      <tm-notifications-table *ngIf = "notificationsTableRowModels.length > 0"
                              [guessTimezone]="guessTimezone"
                              [notificationsTableRowModels]="notificationsTableRowModels"
                              [notificationsTableRowModelsSortBy]="notificationsTableRowModelsSortBy"
                              [notificationsTableRowModelsSortOrder]="notificationsTableRowModelsSortOrder"
                              (deleteNotificationEvent)="deleteNotificationHandler($event)"
                              (loadNotificationEditFormEvent)="loadNotificationEditFormHandler($event)"
                              (sortNotificationsTableRowModelsEvent)="sortNotificationsTableRowModelsHandler($event)"></tm-notifications-table>
    </div>
  </div>
  <br/>
</tm-loading-retry>
