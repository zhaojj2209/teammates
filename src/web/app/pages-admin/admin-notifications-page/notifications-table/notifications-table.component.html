<div class="row">
  <div class="col-12">
    <table id="notifications-table" class="table table-responsive-lg table-striped table-bordered margin-bottom-0">
      <thead>
      <tr [ngClass]="{ 'bg-primary text-white': headerColorScheme === NotificationsTableHeaderColorScheme.BLUE }">
        <th class="sortable-header" (click)="sortNotificationsTableRowModels(SortBy.NOTIFICATION_TITLE)">
          Title
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <i class="fas fa-sort-down" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_TITLE && notificationsTableRowModelsSortOrder === SortOrder.DESC"></i>
            <i class="fas fa-sort-up" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_TITLE && notificationsTableRowModelsSortOrder === SortOrder.ASC"></i>
          </span>
        </th>
        <th class="sortable-header" (click)="sortNotificationsTableRowModels(SortBy.NOTIFICATION_START_TIME)">
          Start Time
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <i class="fas fa-sort-down" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_START_TIME && notificationsTableRowModelsSortOrder === SortOrder.DESC"></i>
            <i class="fas fa-sort-up" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_START_TIME && notificationsTableRowModelsSortOrder === SortOrder.ASC"></i>
          </span>
        </th>
        <th class="sortable-header" (click)="sortNotificationsTableRowModels(SortBy.NOTIFICATION_END_TIME)">
          End Time
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <i class="fas fa-sort-down" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_END_TIME && notificationsTableRowModelsSortOrder === SortOrder.DESC"></i>
            <i class="fas fa-sort-up" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_END_TIME && notificationsTableRowModelsSortOrder === SortOrder.ASC"></i>
          </span>
        </th>
        <th class="sortable-header" (click)="sortNotificationsTableRowModels(SortBy.NOTIFICATION_TARGET_USER)">
          Target User
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <i class="fas fa-sort-down" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_TARGET_USER && notificationsTableRowModelsSortOrder === SortOrder.DESC"></i>
            <i class="fas fa-sort-up" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_TARGET_USER && notificationsTableRowModelsSortOrder === SortOrder.ASC"></i>
          </span>
        </th>
        <th class="sortable-header" (click)="sortNotificationsTableRowModels(SortBy.NOTIFICATION_STYLE)">
          Style
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <i class="fas fa-sort-down" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_STYLE && notificationsTableRowModelsSortOrder === SortOrder.DESC"></i>
            <i class="fas fa-sort-up" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_STYLE && notificationsTableRowModelsSortOrder === SortOrder.ASC"></i>
          </span>
        </th>
        <th class="sortable-header" (click)="sortNotificationsTableRowModels(SortBy.NOTIFICATION_CREATE_TIME)">
          Creation Time
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <i class="fas fa-sort-down" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_CREATE_TIME && notificationsTableRowModelsSortOrder === SortOrder.DESC"></i>
            <i class="fas fa-sort-up" *ngIf="notificationsTableRowModelsSortBy === SortBy.NOTIFICATION_CREATE_TIME && notificationsTableRowModelsSortOrder === SortOrder.ASC"></i>
          </span>
        </th>
        <th>Actions</th>
      </tr>
      </thead>
      <tbody>
      <tr *ngFor="let notificationsTableRowModel of notificationsTableRowModels" [id]="notificationsTableRowModel.notification.notificationId" [ngClass]="{ 'table-success': notificationsTableRowModel.isHighlighted }">
        <td class="text-break">{{ notificationsTableRowModel.notification.title }}</td>
        <td class="text-break"><span class="ngb-tooltip-class" ngbTooltip="{{ notificationsTableRowModel.notification.startTimestamp | formatDateDetail: guessTimezone }}">{{ notificationsTableRowModel.notification.startTimestamp | formatDateBrief: guessTimezone }}</span></td>
        <td class="text-break"><span class="ngb-tooltip-class" ngbTooltip="{{ notificationsTableRowModel.notification.endTimestamp | formatDateDetail: guessTimezone }}">{{ notificationsTableRowModel.notification.endTimestamp | formatDateBrief: guessTimezone }}</span></td>
        <td class="text-break">{{ notificationsTableRowModel.notification.targetUser }}</td>
        <td class="text-break" [ngClass]="notificationsTableRowModel.notification.style | notificationStyleClass">{{ notificationsTableRowModel.notification.style | notificationStyleDescription }}</td>
        <td class="text-break"><span class="ngb-tooltip-class" ngbTooltip="{{ notificationsTableRowModel.notification.createdAt | formatDateDetail: guessTimezone }}">{{ notificationsTableRowModel.notification.createdAt | formatDateBrief: guessTimezone }}</span></td>
        <td class="actions-cell">
          <div class="d-flex">
            <button type="button" class="btn btn-light btn-sm" (click)="loadNotificationEditForm(notificationsTableRowModel.notification)">
              Edit
            </button>
            <button type="button" class="btn btn-danger btn-sm" (click)="deleteNotification(notificationsTableRowModel.notification.notificationId, notificationsTableRowModel.notification.title)">
              Delete
            </button>
          </div>
        </td>
      </tr>
      </tbody>
    </table>
  </div>
</div>
