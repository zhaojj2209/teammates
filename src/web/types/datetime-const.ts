export enum Hours {
    ZERO = 0,
    TWELVE = 12,
    IN_ONE_DAY = 24,
    IN_THREE_DAYS = 72,
    IN_ONE_WEEK = 168,
}

export enum Seconds {
    IN_ONE_MINUTE = 60,
}

export enum Minutes {
    IN_ONE_HOUR = 60,
}

export enum Milliseconds {
    IN_ONE_SECOND = 1000,
    IN_TEN_SECONDS = 10000,
    IN_ONE_MINUTE = 60000,
    IN_TEN_MINUTES = 600000,
    IN_FIFTEEN_MINUTES = 900000,
    IN_ONE_HOUR = 3600000,
    IN_ONE_DAY = 86400000,
    IN_ONE_WEEK = 604800000,
}

export const getDefaultDateFormat = (): DateFormat => {
    return { year: 0, month: 0, day: 0 };
};

export const getLatestTimeFormat = (): TimeFormat => {
    return { hour: 23, minute: 59 };
};

export const getDefaultTimeFormat = (): TimeFormat => {
    return { hour: 0, minute: 0 };
};

export interface DateFormat {
    year: number;
    month: number;
    day: number;
}

export interface TimeFormat {
    hour: number;
    minute: number;
}
