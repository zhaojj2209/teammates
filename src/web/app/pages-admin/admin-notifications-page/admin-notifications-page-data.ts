import { Notification, NotificationStyle, NotificationTargetUser } from '../../../types/api-output';
import { NotificationEditFormModel } from './notification-edit-form/notification-edit-form-model';

/**
 * Structure for example of notification edit model.
 */
export const EXAMPLE_NOTIFICATION_EDIT_MODEL: NotificationEditFormModel = {
  notificationId: 'notification1',
  shown: false,

  startTime: { hour: 0, minute: 0 },
  startDate: { year: 0, month: 0, day: 0 },
  endTime: { hour: 0, minute: 0 },
  endDate: { year: 0, month: 0, day: 0 },

  style: NotificationStyle.SUCCESS,
  targetUser: NotificationTargetUser.INSTRUCTOR,

  title: 'valid title',
  message: 'valid message',

  isSaving: false,
  isDeleting: false,
};
/**
 * Structure for example of notification.
 */
export const EXAMPLE_NOTIFICATION_ONE: Notification = {
  notificationId: 'notification1',
  startTimestamp: 0,
  endTimestamp: 1,
  createdAt: 0,
  updatedAt: 0,
  style: NotificationStyle.SUCCESS,
  targetUser: NotificationTargetUser.INSTRUCTOR,
  title: 'valid title',
  message: 'valid message',
  shown: false,
};
/**
 * Structure for example of notification.
 */
export const EXAMPLE_NOTIFICATION_TWO: Notification = {
  notificationId: 'notification2',
  startTimestamp: 1554960204,
  endTimestamp: 1554977204,
  createdAt: 1554232400,
  updatedAt: 1554232400,
  style: NotificationStyle.DANGER,
  targetUser: NotificationTargetUser.GENERAL,
  title: 'valid title 2',
  message: 'valid message 2',
  shown: false,
};
