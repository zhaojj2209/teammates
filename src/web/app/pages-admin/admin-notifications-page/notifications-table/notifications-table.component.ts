import { Component, EventEmitter, Input, Output } from '@angular/core';
import { SimpleModalService } from '../../../../services/simple-modal.service';
import { Notification } from '../../../../types/api-output';
import { SortBy, SortOrder } from '../../../../types/sort-properties';
import { SimpleModalType } from '../../../components/simple-modal/simple-modal-type';
import { NotificationsTableHeaderColorScheme, NotificationsTableRowModel } from './notifications-table-model';

@Component({
  selector: 'tm-notifications-table',
  templateUrl: './notifications-table.component.html',
  styleUrls: ['./notifications-table.component.scss'],
})
export class NotificationsTableComponent {
  SortBy = SortBy;
  SortOrder = SortOrder;
  NotificationsTableHeaderColorScheme = NotificationsTableHeaderColorScheme;

  @Input()
  guessTimezone = 'UTC';

  @Input()
  notificationsTableRowModels: NotificationsTableRowModel[] = [];

  @Input()
  notificationsTableRowModelsSortBy = SortBy.NOTIFICATION_CREATE_TIME;

  @Input()
  notificationsTableRowModelsSortOrder = SortOrder.DESC;

  @Input()
  headerColorScheme = NotificationsTableHeaderColorScheme.BLUE;

  @Output()
  sortNotificationsTableRowModelsEvent: EventEmitter<SortBy> = new EventEmitter();

  @Output()
  deleteNotificationEvent: EventEmitter<String> = new EventEmitter();

  @Output()
  loadNotificationEditFormEvent: EventEmitter<Notification> = new EventEmitter();

  constructor(private simpleModalService: SimpleModalService) { }

  /**
   * Sorts the list of feedback session row.
   */
  sortNotificationsTableRowModels(by: SortBy): void {
    this.sortNotificationsTableRowModelsEvent.emit(by);
  }

  /**
   * Deletes a notification based on its ID.
   */
  deleteNotification(notificationId: string, title: string): void {
    const modalRef = this.simpleModalService.openConfirmationModal(
      'Confirm your action',
      SimpleModalType.DANGER,
      `Do you want to delete this notification (titled "${title}") permanently? This action will not be reversible.`,
    );
    modalRef.result.then(() => this.deleteNotificationEvent.emit(notificationId));
  }

  /**
   * Loads the notification edit form.
   */
  loadNotificationEditForm(notification: Notification): void {
    this.loadNotificationEditFormEvent.emit(notification);
  }
}
