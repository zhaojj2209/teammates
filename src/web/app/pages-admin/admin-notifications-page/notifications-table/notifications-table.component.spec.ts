import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import moment from 'moment-timezone';
import { SortBy, SortOrder } from '../../../../types/sort-properties';
import { EXAMPLE_NOTIFICATION_ONE, EXAMPLE_NOTIFICATION_TWO } from '../admin-notifications-page-data';
import { AdminNotificationsPageModule } from '../admin-notifications-page.module';
import { NotificationsTableRowModel } from './notifications-table-model';

import { NotificationsTableComponent } from './notifications-table.component';

describe('NotificationsTableComponent', () => {
  let component: NotificationsTableComponent;
  let fixture: ComponentFixture<NotificationsTableComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        AdminNotificationsPageModule,
        HttpClientTestingModule,
      ],
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(NotificationsTableComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  const notificationTableRowModel1: NotificationsTableRowModel = {
    isHighlighted: true,
    notification: EXAMPLE_NOTIFICATION_ONE,
  };

  const notificationTableRowModel2: NotificationsTableRowModel = {
    isHighlighted: false,
    notification: EXAMPLE_NOTIFICATION_TWO,
  };

  it('should snap like in notification page with 2 notifications sorted by create time', () => {
    moment.tz.setDefault('UTC');
    component.notificationsTableRowModels = [notificationTableRowModel1, notificationTableRowModel2];
    component.notificationsTableRowModelsSortBy = SortBy.NOTIFICATION_CREATE_TIME;
    component.notificationsTableRowModelsSortOrder = SortOrder.DESC;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap like in notification page with 2 notifications sorted by start time', () => {
    moment.tz.setDefault('UTC');
    component.notificationsTableRowModels = [notificationTableRowModel1, notificationTableRowModel2];
    component.notificationsTableRowModelsSortBy = SortBy.NOTIFICATION_START_TIME;
    component.notificationsTableRowModelsSortOrder = SortOrder.DESC;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
