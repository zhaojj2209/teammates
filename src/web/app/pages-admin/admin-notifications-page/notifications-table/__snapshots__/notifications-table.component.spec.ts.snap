// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsTableComponent should snap like in notification page with 2 notifications sorted by create time 1`] = `
<tm-notifications-table
  NotificationsTableHeaderColorScheme={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  deleteNotificationEvent={[Function EventEmitter_]}
  guessTimezone={[Function String]}
  headerColorScheme="0"
  loadNotificationEditFormEvent={[Function EventEmitter_]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  simpleModalService={[Function SimpleModalService]}
  sortNotificationsTableRowModelsEvent={[Function EventEmitter_]}
  timezoneService={[Function TimezoneService]}
>
  <div
    class="row"
  >
    <div
      class="col-12"
    >
      <table
        class="table table-responsive-lg table-striped table-bordered margin-bottom-0"
        id="notifications-table"
      >
        <thead>
          <tr
            class="bg-primary text-white"
          >
            <th
              class="sortable-header"
            >
               Title 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Start Time 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               End Time 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Target User 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Style 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Creation Time 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
                <i
                  class="fas fa-sort-down"
                />
              </span>
            </th>
            <th>
              Actions
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="table-success"
          >
            <td
              class="text-break"
            >
              valid title
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                1 Jan 7:30 AM
              </span>
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                1 Jan 7:30 AM
              </span>
            </td>
            <td
              class="text-break"
            >
              INSTRUCTOR
            </td>
            <td
              class="text-break bg-success text-white"
            >
              Success (green)
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                1 Jan 7:30 AM
              </span>
            </td>
            <td
              class="actions-cell"
            >
              <div
                class="d-flex"
              >
                <button
                  class="btn btn-light btn-sm"
                  type="button"
                >
                   Edit 
                </button>
                <button
                  class="btn btn-danger btn-sm"
                  type="button"
                >
                   Delete 
                </button>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="text-break"
            >
              valid title 2
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                19 Jan 7:26 AM
              </span>
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                19 Jan 7:26 AM
              </span>
            </td>
            <td
              class="text-break"
            >
              GENERAL
            </td>
            <td
              class="text-break bg-danger text-white"
            >
              Danger (red)
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                19 Jan 7:13 AM
              </span>
            </td>
            <td
              class="actions-cell"
            >
              <div
                class="d-flex"
              >
                <button
                  class="btn btn-light btn-sm"
                  type="button"
                >
                   Edit 
                </button>
                <button
                  class="btn btn-danger btn-sm"
                  type="button"
                >
                   Delete 
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</tm-notifications-table>
`;

exports[`NotificationsTableComponent should snap like in notification page with 2 notifications sorted by start time 1`] = `
<tm-notifications-table
  NotificationsTableHeaderColorScheme={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  deleteNotificationEvent={[Function EventEmitter_]}
  guessTimezone={[Function String]}
  headerColorScheme="0"
  loadNotificationEditFormEvent={[Function EventEmitter_]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  simpleModalService={[Function SimpleModalService]}
  sortNotificationsTableRowModelsEvent={[Function EventEmitter_]}
  timezoneService={[Function TimezoneService]}
>
  <div
    class="row"
  >
    <div
      class="col-12"
    >
      <table
        class="table table-responsive-lg table-striped table-bordered margin-bottom-0"
        id="notifications-table"
      >
        <thead>
          <tr
            class="bg-primary text-white"
          >
            <th
              class="sortable-header"
            >
               Title 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Start Time 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
                <i
                  class="fas fa-sort-down"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               End Time 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Target User 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Style 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Creation Time 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th>
              Actions
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="table-success"
          >
            <td
              class="text-break"
            >
              valid title
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                1 Jan 7:30 AM
              </span>
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                1 Jan 7:30 AM
              </span>
            </td>
            <td
              class="text-break"
            >
              INSTRUCTOR
            </td>
            <td
              class="text-break bg-success text-white"
            >
              Success (green)
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                1 Jan 7:30 AM
              </span>
            </td>
            <td
              class="actions-cell"
            >
              <div
                class="d-flex"
              >
                <button
                  class="btn btn-light btn-sm"
                  type="button"
                >
                   Edit 
                </button>
                <button
                  class="btn btn-danger btn-sm"
                  type="button"
                >
                   Delete 
                </button>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="text-break"
            >
              valid title 2
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                19 Jan 7:26 AM
              </span>
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                19 Jan 7:26 AM
              </span>
            </td>
            <td
              class="text-break"
            >
              GENERAL
            </td>
            <td
              class="text-break bg-danger text-white"
            >
              Danger (red)
            </td>
            <td
              class="text-break"
            >
              <span
                class="ngb-tooltip-class"
              >
                19 Jan 7:13 AM
              </span>
            </td>
            <td
              class="actions-cell"
            >
              <div
                class="d-flex"
              >
                <button
                  class="btn btn-light btn-sm"
                  type="button"
                >
                   Edit 
                </button>
                <button
                  class="btn btn-danger btn-sm"
                  type="button"
                >
                   Delete 
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</tm-notifications-table>
`;

exports[`NotificationsTableComponent should snap with default fields 1`] = `
<tm-notifications-table
  NotificationsTableHeaderColorScheme={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  deleteNotificationEvent={[Function EventEmitter_]}
  guessTimezone={[Function String]}
  headerColorScheme="0"
  loadNotificationEditFormEvent={[Function EventEmitter_]}
  notificationsTableRowModels={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  simpleModalService={[Function SimpleModalService]}
  sortNotificationsTableRowModelsEvent={[Function EventEmitter_]}
  timezoneService={[Function TimezoneService]}
>
  <div
    class="row"
  >
    <div
      class="col-12"
    >
      <table
        class="table table-responsive-lg table-striped table-bordered margin-bottom-0"
        id="notifications-table"
      >
        <thead>
          <tr
            class="bg-primary text-white"
          >
            <th
              class="sortable-header"
            >
               Title 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Start Time 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               End Time 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Target User 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Style 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </th>
            <th
              class="sortable-header"
            >
               Creation Time 
              <span
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
                <i
                  class="fas fa-sort-down"
                />
              </span>
            </th>
            <th>
              Actions
            </th>
          </tr>
        </thead>
        <tbody>
        </tbody>
      </table>
    </div>
  </div>
</tm-notifications-table>
`;
